### PR TITLE
implementation of the twitch access token, and get user info by login…

### LINKED ISSUE #130: [FEATURE]: Twitch Service
### Description

Connecting to the Twitch API, performing OAuth2 authentication, then performing reaction actions in relation to the API.

### To do

implementation of the Twitch oAuth.
Implementations of all these Actions :
Get if he is on live
Get if it has more than 1k viewers
Get if a game is the top twitch game

Implementations of all these Reactions:
Run a poll
Send live info
Send Whisper

### Priority

Medium

### Project parts

- [X] Server
- [X] Application
- [X] Web
- [ ] Automation

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by STom6**
Connecting to the Twitch API, performing OAuth2 authentication, then performing reaction actions in relation to the API.
# New feature
implementation of the Twitch oAuth.
Implementations of all these Actions :
Get if he is on live
Get if it has more than selected spectators
Get if a game is the top twitch game

# Bug fixed

### CODE REVIEW

File: Server/package.json [hunk 1]
```diff
@@ -13,6 +13,9 @@
     "login-with-twitter": "^1.2.4",
     "node-cron": "^3.0.2",
     "nodemailer": "^6.9.0",
+    "twitch": "^4.6.7",
+    "twitch-api": "^0.5.0",
```
Comment (31Nathan) [inline Server/package.json]: All are necessary?

File: Server/Services/TwitchService.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    "analytics:read:games",
+    "moderator:read:followers",
+    "channel:manage:moderators",
+    "channel:manage:predictions",
+    "channel:manage:polls",
+    "user:manage:whispers"
+].join(" ");
+
+/**
+ * Required variable use in authentification
```
Comment (osvegn) [inline Server/Services/TwitchService.js]: used

File: Server/Services/TwitchService.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+ */
+function encodeQueryString(params) {
+    const queryString = new URLSearchParams();
+    for (let paramName in params) {
+        queryString.append(paramName, params[paramName]);
+    }
+    return queryString.toString();
+}
+
+function encodeUrlScope(params)
```
Comment (osvegn) [inline Server/Services/TwitchService.js]: undocumented

File: Server/app.js [hunk 4]
```diff
@@ -43,22 +46,22 @@ app.get("/testConnexion", (req, res) => {
     res.send("Connexion established").status(200);
 })
 
-nodeCron.schedule("*/10 * * * * *", () => {
-    try {
-        firebaseFunctions.getAllUsersFromFireBase().then(data => {
+// nodeCron.schedule("*/10 * * * * *", () => {
```
Comment (osvegn) [inline Server/app.js]: Write a comment to let know why it is commented

**Review by osvegn — CHANGES_REQUESTED** [review]
Some documentation is missing, otherwise it seems good.

File: Application/src/Screens/SettingsScreen.tsx [hunk 5]
```diff
@@ -3,6 +3,8 @@ import { StyleSheet, SafeAreaView, View, Image, Text, TouchableOpacity, ImageSou
 import { Globals } from "../Common/Globals";
 import Geolocation from 'react-native-geolocation-service';
 import { SettingsProps } from "../Common/Interfaces";
+import { firebase } from "@react-native-firebase/auth";
+import { locationURL } from 'location-href';
```
Comment (JohanCDev) [inline Application/src/Screens/SettingsScreen.tsx]: unused

File: Application/package.json [hunk 6]
```diff
@@ -20,7 +20,8 @@
     "react-native-fbsdk-next": "^11.1.0",
     "react-native-geolocation-service": "^5.3.1",
     "react-native-navigation": "^7.31.0",
-    "react-native-uuid": "^2.0.1"
+    "react-native-uuid": "^2.0.1",
+    "location-href" : "^1.0.0"
```
Comment (JohanCDev) [inline Application/package.json]: Unused

**Review by JohanCDev — DISMISSED** [review]

File: Application/package.json (hunk 6 above)
Comment (STom6) [reply Application/package.json]: Done

File: Application/src/Screens/SettingsScreen.tsx (hunk 5 above)
Comment (STom6) [reply Application/src/Screens/SettingsScreen.tsx]: Done

**Comment by JohanCDev** [comment]
Missing Web, see with @osvegn

**Review by CedricCORGE — DISMISSED** [review]
LGTM

File: Application/src/Screens/SettingsScreen.tsx [hunk 7]
```diff
@@ -39,7 +39,34 @@ export default function SettingsScreen(props: SettingsProps) {
         getIpStatus()
         // getAddressFromCoordinates(props.userInfo.coord.latitude, props.userInfo.coord.longitude)
     }, [])
+    /**
+     * Gets the elements of the url and encodes them to work with "éàè..." characters
+     * @function encodeQueryString
+     * @param {*} params contains the elements to be added to the url.
+     */
```
Comment (osvegn) [inline Application/src/Screens/SettingsScreen.tsx:47]: Missing return field

File: Application/src/Screens/SettingsScreen.tsx [hunk 8]
```diff
@@ -39,7 +39,34 @@ export default function SettingsScreen(props: SettingsProps) {
...
+            queryString.append(paramName, params[paramName]);
+        }
+        return queryString.toString();
+    }
 
+    /**
+     * Encode Uri 
+     * @function encodeQueryString
+     * @param {*} params contains the elements to be added to the url.
+     */
```
Comment (osvegn) [inline Application/src/Screens/SettingsScreen.tsx:61]: Missing return field

File: Server/Services/TwitchService.js [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    return queryString.toString();
+}
+
+/**
+ * Check if the Top Twich game is the same as the one selected by the user
+ * @function checkTopGames
+ * @param {String} uid uid of the user
+ * @param {String} game Game selected by the user
+ */
```
Comment (osvegn) [inline Server/Services/TwitchService.js:32]: Missing return field

File: Server/Services/TwitchService.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+ * @constant firebaseFunctions
+ * @requires firebaseFunctions
+ */
+const firebaseFunctions = require('../firebaseFunctions')
+
+/**
+ * Gets the elements of the url and encodes them to work with "éàè..." characters
+ * @function encodeQueryString
+ * @param {*} params params contains the elements to be added to the url.
+ */
```
Comment (osvegn) [inline Server/Services/TwitchService.js:17]: Missing return field

File: Server/Services/TwitchService.js [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    })
+    .catch(error => {
+        console.log(error);
+    });
+}
+
+module.exports = {
+    /**
+     * function tree which allows you to choose the right function
+     * @param {string} func function chosen by the user
```
Comment (osvegn) [inline Server/Services/TwitchService.js:136]: missing function field

File: Server/app.js [hunk 12]
```diff
@@ -14,19 +15,62 @@ const port = config.port;
...
+    state: true
+  },
+  function(accessToken, refreshToken, profile, done) {
+    profile.accessToken = accessToken;
+    profile.refreshToken = refreshToken;
+
+    // Securely store user profile in your DB
+    //User.findOrCreate(..., function(err, user) {
+    //  done(err, user);
+    //});
```
Comment (osvegn) [inline Server/app.js]: Useful ?

File: Server/app.js [hunk 13]
```diff
@@ -156,13 +200,46 @@ app.get('/register/iss', (req, res) => {
...
+
+app.get('/twitch/doAct', (req, res) => {
+    TwitchService.actionTwitch("stream", firebaseUid, "krl_stream")
+    res.send("TWITCH ACT");
 })
 
-app.get('/tw', (req, res) => {
-    twitterService.ActionTw('like', 'chelsea', firebaseUid, req, res)
+app.get('/twitter', (req, res) => {
+    res.render('index')
```
Comment (osvegn) [inline Server/app.js:235]: Why is it updated ?

File: Web/src/SettingsPage.js [hunk 14]
```diff
@@ -228,57 +228,54 @@ export default function SettingsPage(props) {
 
 
     /**
-     * Ensure the log in of the user in the services
-     * @function LogWith
-     * @param {string} serviceName the name of the service where you want to register
+     * Ensure the log in of the user on Spotify
+     * @function spotifyConnexion
```
Comment (osvegn) [inline Web/src/SettingsPage.js:232]: missing async field (it has to be empty)

File: Web/src/SettingsPage.js [hunk 15]
```diff
@@ -289,13 +286,95 @@ export default function SettingsPage(props) {
...
                 <img src={props.image} style={styles.serviceImage}></img>
                 <p style={styles.serviceText}>Connexion à {props.service}</p>
                 <img src={ArrowRight} style={styles.serviceArrow}></img>
             </div>
         )
     }
+
+    /**
+     * Authenticates the user with Twitch API.
+     * @async
```
Comment (osvegn) [inline Web/src/SettingsPage.js]: It's not an async function

File: Web/src/SettingsPage.js [hunk 16]
```diff
@@ -289,13 +286,95 @@ export default function SettingsPage(props) {
...
+        const response_type = "token"
+
+        twitchAuth(scopes, twitch_oauth_url, response_type)
+    }
+
+    /**
+     * Encode Uri
+     * @function encodeQueryString
+     * @param {*} params contains the elements to be added to the url.
+     */
```
Comment (osvegn) [inline Web/src/SettingsPage.js:323]: missing return field

**Review by osvegn — DISMISSED** [review]
Some missing documentation or useless things

File: Server/app.js (hunk 13 above)
Comment (STom6) [reply Server/app.js:235]: unnecessary

File: Web/src/SettingsPage.js (hunk 15 above)
Comment (STom6) [reply Web/src/SettingsPage.js]: done

File: Web/src/SettingsPage.js (hunk 16 above)
Comment (STom6) [reply Web/src/SettingsPage.js:323]: done

File: Web/src/SettingsPage.js (hunk 14 above)
Comment (STom6) [reply Web/src/SettingsPage.js:232]: done

File: Server/Services/TwitchService.js (hunk 11 above)
Comment (STom6) [reply Server/Services/TwitchService.js:136]: done

File: Server/Services/TwitchService.js (hunk 9 above)
Comment (STom6) [reply Server/Services/TwitchService.js:32]: done

File: Server/Services/TwitchService.js (hunk 10 above)
Comment (STom6) [reply Server/Services/TwitchService.js:17]: done

File: Application/src/Screens/SettingsScreen.tsx (hunk 8 above)
Comment (STom6) [reply Application/src/Screens/SettingsScreen.tsx:61]: done

File: Application/src/Screens/SettingsScreen.tsx (hunk 7 above)
Comment (STom6) [reply Application/src/Screens/SettingsScreen.tsx:47]: done

**Review by CedricCORGE — DISMISSED** [review]
lgtm

**Review by 31Nathan — APPROVED** [review]
LGTM

**Review by CedricCORGE — APPROVED** [review]
LGTm

**Pull request merged**